### PR TITLE
fix(gatsby-source-contentful): fixed contentful asset download stalling with high number of assets

### PR DESCRIPTION
**Pull request by me4502**
## Description

With large Contentful sites, the number of assets can open too many connections for Node to handle. This leads to the download stalling. While the issue is unlikely to be encountered on a CI instance, on a local machine it can freeze downloads or even crash the machine in some cases. This PR adds a workload distribution system that limits the number of concurrent downloads, fixing the issue. The number is also configurable using a plugin option.

### Documentation

Docs in the readme are updated.

### CODE REVIEW

**Review by pvdz — APPROVED** [review]
I really like this approach. And you're right, we needed this throttle.

Thank you! :purple_heart:

**Comment by pvdz** [comment]
Oh, well some tests need to be updated I guess.

File: packages/gatsby-source-contentful/src/download-contentful-assets.js [hunk 1]
```diff
@@ -9,7 +9,23 @@ const bar = new ProgressBar(
...
+ * @param workers A list of async functions to complete
+ * @param {number} count The number of task runners to use
+ */
+
+async function distributeWorkload(workers, count) {
+  const methods = workers.slice()
+
+  async function task() {
+    while (methods.length > 0) {
+      await methods.shift()
```
Comment (pvdz) [inline packages/gatsby-source-contentful/src/download-contentful-assets.js:23]: While a change is necessary anyways; are you sure `.pop()` wouldn't be faster for really large arrays (200k+)? Internally it _ought_ to just move the offset pointer so I think it's equal.

Comment (me4502) [reply packages/gatsby-source-contentful/src/download-contentful-assets.js]: Good catch, just switched it. The code I adapted this from tried to keep some sense of order in the returned results so hence used shift.

**Review by pvdz — APPROVED** [review]
Awesome! Thank you :purple_heart:

**Comment by pvdz** [comment]
Published in `gatsby-source-contentful@3.0.2` (same time as `gatsby@2.24.84` although this is a standalone fix)

**Comment by pvdz** [comment]
@me4502 this seems to be "breaking" the progress bar of the reporter because now it's printing one 1/1 line per image downloaded. Also _appears_ to be doing it one by one, rather than in pools of 50. Are you seeing this as well?

**Comment by me4502** [comment]
I'll take a look tomorrow (Australia time), there's a chance some of the minor cleanups I did at the end may have modified behaviour - but it's working fine on my local patch.

It does increase the number over time as it hasn't yet triggered the remote file download yet, but in my testing it was always faster to increase the number than it was to download, so it never triggered multiple messages (unless it was set to a small number such as 2)

**Comment by pvdz** [comment]
I got it. The `count` needs to have a default in this function as well because it's receiving `undefined`, causing the array to schedule just one worker :) I'll fix it.

**Comment by me4502** [comment]
Ah that explains it - I removed that when I added the config value as I assumed Joi would handle it. Thanks :)

**Pull request merged**